### PR TITLE
Add an option to configure site constants

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Replaces #12457 where it was expected to be able to update `SCRIPT_DEBUG` constant:

> When I setup Gutenberg for local development, I'd expect it to load unminified files.

Takes out part of #13452 where it is necessary to be able to set `WP_DEBUG` constant.

This PR enables an option to make debugging WordPress easier with the existing Docker setup by providing env variables which reflect the same PHP constants used in WordPress:

* `WP_DEBUG`
* `SCRIPT_DEBUG`

See more at https://codex.wordpress.org/Debugging_in_WordPress.

## Testing

To enable all available site constant variables run the following:
```sh
SCRIPT_DEBUG=true WP_DEBUG=true ./bin/setup-local-env.sh
```

To disable all available site constant variables run the following:
```sh
SCRIPT_DEBUG=false WP_DEBUG=false ./bin/setup-local-env.sh
```

### CODE REVIEW

**Comment by gziolo** [comment]
Yay, it works https://travis-ci.com/WordPress/gutenberg/jobs/183804711#L978-L978.

Thanks @aduth and @pento for inspiration how to tackle it :)

I will polish PR real quick and add missing description and docs.

File: .travis.yml [hunk 1]
```diff
@@ -65,7 +65,7 @@ jobs:
         - ./bin/run-wp-unit-tests.sh
 
     - name: E2E tests (Admin with plugins) (1/2)
-      env: WP_VERSION=latest POPULAR_PLUGINS=true
+      env: WP_VERSION=latest POPULAR_PLUGINS=true WP_DEBUG=true WP_DEBUG_DISPLAY=true SCRIPT_DEBUG=true
```
Comment (gziolo) [inline .travis.yml]: I will remove this change as soon as Docker run is finished and I will confirm that it also works with Travis.

Comment (gziolo) [reply .travis.yml]: See f949a17.

File: bin/install-wordpress.sh [hunk 2]
```diff
@@ -90,6 +90,24 @@ if [ "$CURRENT_URL" != "http://localhost:$HOST_PORT" ]; then
 	docker-compose $DOCKER_COMPOSE_FILE_OPTIONS run --rm -u 33 $CLI option update siteurl "http://localhost:$HOST_PORT" --quiet
 fi
 
+# Configure site constants.
+echo -e $(status_message "Configuring site constants...")
```
Comment (gziolo) [inline bin/install-wordpress.sh]: There is a lot of code duplication here. Commits which simplify it are warmly welcomed :)

**Comment by ellatrix** [comment]
Why don't we just enable them by default?

**Comment by gziolo** [comment]
e2e tests will fail, there are still React warnings coming from class methods which are marked as deprecated in the strict mode :(

**Comment by ellatrix** [comment]
Maybe we can add an alternative command that includes these, in package.json?

**Comment by ellatrix** [comment]
Just saying that when I'm setting up Gutenberg locally, I use it for development purposes. Sounds like we'd benefit from a less verbose setup command.

**Comment by gziolo** [comment]
Feel free to add some changes to this PR if you have some ideas. I just wanted to be able to use `SCRIPT_DEBUG` myself :) In addition, @aduth wants to enable `WP_DEBUG` so we could catch PHP errors produced by plugin.

**Comment by aduth** [comment]
> e2e tests will fail, there are still React warnings coming from class methods which are marked as deprecated in the strict mode :(

Is it reasonable to just override those temporarily for the end-to-end tests, to disable the script debug?

**Comment by gziolo** [comment]
> Is it reasonable to just override those temporarily for the end-to-end tests, to disable the script debug?

I think it makes a lot of sense. It didn't occurred to me before :)

File: bin/install-wordpress.sh (hunk 2 above)
Comment (mcsf) [reply bin/install-wordpress.sh]: We could use macro-like behaviour with the following:

```bash
$ A=hello; B=salut
$ VAR=A; eval echo \$$VAR
hello
$ VAR=B; eval echo \$$VAR
salut
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -79,6 +79,11 @@ Alternatively, you can use your own local WordPress environment and clone this r
 
 Next, open a terminal (or if on Windows, a command prompt) and navigate to the repository you cloned. Now type `npm install` to get the dependencies all set up. Then you can type `npm run dev` in your terminal or command prompt to keep the plugin building in the background as you work on it.
 
+WordPress comes with specific [debug systems](https://codex.wordpress.org/Debugging_in_WordPress) designed to simplify the process as well as standardize code across the core, plugins and themes. It is possible to use environment variables (`WP_DEBUG`, `WP_DEBUG_DISPLAY` and `SCRIPT_DEBUG`) to update site's config constants located in `wp-config.php` file. They can be updated at any time by running the following command:
```
Comment (mcsf) [inline CONTRIBUTING.md]:
```suggestion
WordPress comes with specific [debug systems](https://codex.wordpress.org/Debugging_in_WordPress) designed to simplify the process as well as standardize code across core, plugins and themes. It is possible to use environment variables (`WP_DEBUG`, `WP_DEBUG_DISPLAY` and `SCRIPT_DEBUG`) to update a site's configuration constants located in `wp-config.php` file. They can be updated at any time by running the following command:
```

**Comment by gziolo** [comment]
I'm seeing the following on Travis:
```
Error: The constant 'WP_DEBUG_DISPLAY' is not defined in the 'wp-config.php' file.
The command "./bin/setup-local-env.sh" failed and exited with 1 during .
```

I added a commit which removed the change for `WP_DEBUG_DISPLAY`: 0eb93b6. It looks like this is a limitation of WP CLI as of today:
https://stackoverflow.com/questions/35711003/wp-cli-toggle-wp-debug/49105953#comment91168718_49105953

**Comment by gziolo** [comment]
Hmmm, it's also being reported for `SCRIPT_DEBUG`:
```
Error: The constant 'SCRIPT_DEBUG' is not defined in the 'wp-config.php' file.
```

File: bin/bootstrap-env.sh [hunk 4]
```diff
@@ -1,6 +1,8 @@
 #!/usr/bin/env bash
 
 WP_VERSION=${WP_VERSION-latest}
+WP_DEBUG=${WP_DEBUG-true}
```
Comment (aduth) [inline bin/bootstrap-env.sh]: Possible thought on the error: Is this the correct syntax for assigning the default? Is the colon `:` not required?

See:

- https://stackoverflow.com/a/2013589
- https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html

Comment (aduth) [reply bin/bootstrap-env.sh]: Also a bit curious why we don't use a `.env` file: https://docs.docker.com/compose/env-file/

Comment (gziolo) [reply bin/bootstrap-env.sh]: I followed other occurrences in the file. I think it works correctly, although I personally don't know which syntax would be preferred.

The issue was caused by the fact that PHP didn't have those constants defined and WP CLI's implementation is limited to updating existing constants ... https://github.com/WordPress/gutenberg/pull/14371/commits/c34d3c027aea9ddd06639bb5a4b7e9b38cc5ce7f fixes it.

**Comment by gziolo** [comment]
It works with the latest changes applied. I think this is an improvement over what we had before so I would be happy if we could land it after some testing.

**Review by aduth — COMMENTED** [review]
This all _looks_ reasonable, but I'm struggling with the previously-noted error in my own environment.

Stopping / removing all existing containers, then running `./bin/setup-local-env.sh`, I see:

```
⇒ docker-compose down
⇒ ./bin/setup-local-env.sh
# ...
Error: The constant 'SCRIPT_DEBUG' is not defined in the 'wp-config.php' file.
```

Am I missing something, or can you think of some reason why the new environment values specified in `docker-compose.yml` might not take effect in my environment (would others expect to encounter similar)?

File: bin/bootstrap-env.sh (hunk 4 above)
Comment (gziolo) [reply bin/bootstrap-env.sh]: Yes, this is complex to explain but correct. Each constant needs to be defined in `wp-config.php` if we want to be able to update it with WP-CLI.

**Comment by gziolo** [comment]
> Am I missing something, or can you think of some reason why the new environment values specified in docker-compose.yml might not take effect in my environment (would others expect to encounter similar)?

No idea to be honest. It looks similar to the error we had on Travis before c34d3c0 was added. I personally restarted Docker to test it locally.

**Review by aduth — APPROVED** [review]
While I'm not able to test it, I strongly suspect my issues are local, and I'd not want to be the reason for this being blocked. Otherwise, everything _appears_ in good order here. I might offer it that another person test it, though am comfortable with it being merged if you're confident about it.

**Comment by gziolo** [comment]
I will restart Docker and try locally again before I merge tomorrow, thanks for your review 💯

**Comment by gziolo** [comment]
I tested locally after system restarted. Everything works as expected.

**Comment by aduth** [comment]
> e2e tests will fail, there are still React warnings coming from class methods which are marked as deprecated in the strict mode :(

I created a follow-up task at #14845

**Comment by gziolo** [comment]
Thanks @aduth!

**Pull request merged**